pallets/msa/src/lib.rs

```diff
@@ -1292,6 +1349,57 @@ impl<T: Config> Pallet<T> {
 		block_number / (T::BlockNumber::from(T::MortalityWindowSize::get())) %
 			T::BlockNumber::from(T::NumberOfBuckets::get())
 	}
+
+	/// Get list of account ids that are associated with a given msa
+	pub fn get_msa_keys_offchain(msa_id: MessageSourceId) -> Vec<T::AccountId> {
+		let msa_keys_res = offchain_storage::get_msa_keys::<MessageSourceId, T::AccountId>(msa_id);
+		if let Err(e) = msa_keys_res {
+			log_err!("Error getting msa keys from offchain storage: {:?}", e);
+			return Vec::new()
+		}
+		msa_keys_res.unwrap()
+	}
+
+	fn process_events(
+		msa_id: MessageSourceId,
+		events: Vec<(T::AccountId, offchain_storage::EventType)>,
+	) {
+		let lock_status = offchain_common::lock(msa_id.encode().as_slice(), || {
+			for (key, event_type) in &events {
+				match event_type {
+					offchain_storage::EventType::Add => {
+						let add_result = offchain_storage::process_msa_key_event::<
+							MessageSourceId,
+							T::AccountId,
+						>(offchain_storage::MSAPublicKeyDataOperation::Add::<
+							MessageSourceId,
+							T::AccountId,
+						>(msa_id, key.clone()));
+						if let Err(e) = add_result {
+							log_err!("Error adding key to offchain storage: {:?}", e);
+						}
+					},
+					offchain_storage::EventType::Remove => {
+						let delete_result = offchain_storage::process_msa_key_event::<
+							MessageSourceId,
+							T::AccountId,
+						>(
+							offchain_storage::MSAPublicKeyDataOperation::Remove::<
+								MessageSourceId,
+								T::AccountId,
+							>(msa_id, key.clone()),
+						);
+						if let Err(e) = delete_result {
+							log_err!("Error deleting key from offchain storage: {:?}", e);
+						}
+					},
+				}
+			}
+		});
+		if let offchain_common::LockStatus::Locked = lock_status {
+			log_err!("Unable to acquire lock, skipping event processing");
+		}
+	}
```

Comment:
So
* acquire lock on MSA to execute a replay of events and 
* do necessary update to offchain storage for specific msa_id
* lock prevents another offchain_worker to try to mutate same msa_id data